### ISSUE TITLE
GL_DEPTH_TEST not functioning

### ISSUE THREAD

**Issue by tobyclh** [glumpy/glumpy#142]
TLDR gl_depth_test is not properly enabled in the gloo-picking example. Would like to see how to make it work :smiley: 

I was checking out the gloo-picking example, and realize GL_DEPTH_TEST is actually not working.
This is not easily observable but if you think about it carefully, since you are rotating a point cloud at its center, let's say your mouse is going upward, the points at half of the cloud (the half that is closer to you) should move the same direction as your mouse, and __should occlude ALL points that is moving at the opposite direction__, which just simply isn't the case right now.

**Comment by rougier** [comment]
I think the problem is related to the fact that each point is a single vertex (GL_POINT) and depth is tested only for the center of each point (because I did not set the frag_depth for each fragment). If you look at the `protein.py` example, you'll see it works as expected when frag depth is set (inside the shader I mean).

**Comment by tobyclh** [comment]
Thanks for the quick response, but I suspect that its actually about gl_readpixel or the framebuffer.
I quickly made a modified version of the gloo-picking example, set the gl_FragDepth and still no luck.
Below is the modified picking example.
```

# -----------------------------------------------------------------------------
# Copyright (c) 2009-2016 Nicolas P. Rougier. All rights reserved.
# Distributed under the (new) BSD License.
# -----------------------------------------------------------------------------
import numpy as np
from glumpy import app, gl, gloo, glm
from glumpy.transforms import Trackball
app.use('qt5')
vertex = """
#include "math/constants.glsl"

uniform float linewidth;
uniform float antialias;

attribute vec4  fg_color;
attribute vec4  bg_color;
attribute float size;
attribute vec3  position;
attribute float id;

varying vec4  v_id;
varying float v_size;
varying vec4  v_fg_color;
varying vec4  v_bg_color;
varying vec4  v_eye_position;
void main (void)
{
    // id to vec3 (alpha must be one or color will be blended
    v_id = vec4 ( mod(floor(id / (256*256)), 256) / 255.0,
                  mod(floor(id /     (256)), 256) / 255.0,
                  mod(floor(id /       (1)), 256) / 255.0,
                  1.0 );
    v_eye_position = <transform.trackball_view> *
                     <transform.trackball_model> *
                     vec4(position,1.0);
    v_size = size;
    v_fg_color = fg_color;
    v_bg_color = bg_color;
    gl_Position = <transform(vec4(position,1.0))>;
    gl_PointSize = M_SQRT2 * size + 2.0 * (linewidth + 1.5*antialias);
}
"""

fragment = """
#include "markers/disc.glsl"
#include "math/constants.glsl"
#include "antialias/antialias.glsl"

uniform float linewidth;
uniform float antialias;

varying vec4  v_id;
varying float v_size;
varying vec4  v_fg_color;
varying vec4  v_bg_color;
varying vec4 v_eye_position;
void main()
{
    vec2 texcoord = gl_PointCoord* 2.0 - vec2(1.0);
    float x = texcoord.x;
    float y = texcoord.y;
    float d = 1.0 - x*x - y*y;
    if (d <= 0.0) discard;
    float z = sqrt(d);


    vec2 P = gl_PointCoord.xy - vec2(0.5,0.5);
    float point_size = M_SQRT2*v_size  + 2 * (linewidth + 1.5*antialias);
    float distance = marker_disc(P*point_size, v_size);
    vec4 color = outline(distance, linewidth, antialias, v_fg_color, v_bg_color);

    vec4 pos = v_eye_position;
    pos.z += point_size*z;
    pos = <transform.trackball_projection> * pos;
    gl_FragDepth = 0.5*(pos.z / pos.w)+0.5;

    gl_FragData[0] = color;
    gl_FragData[1] = v_id;
}
"""

quad_vertex = """
attribute vec2 position;
varying vec2 v_texcoord;
void main(void)
{
    gl_Position = vec4(position,0,1);
    v_texcoord = (position+1.0)/2.0;
}"""

quad_fragment = """
uniform sampler2D color;
varying vec2 v_texcoord;
void main(void)
{
    gl_FragColor = texture2D(color,v_texcoord);
}
"""

window = app.Window(width=1000, height=1000, color=(1,1,1,1))

n = 5000
program = gloo.Program(vertex, fragment, count=n)
program['position'] = 0.75 * np.random.randn(n,3)
program['size']     = np.random.uniform(20,30,n)
program['fg_color'] = 0,0,0,1
program['bg_color'] = np.random.uniform(0.75, 1.00, (n, 4))
program['bg_color'][:,3] = 1
program['linewidth'] = 1.0
program['antialias'] = 1.0
program['transform'] = Trackball()
program["id"] = np.arange(n,dtype=np.float32)

quad = gloo.Program(quad_vertex, quad_fragment, count=4)
quad['position']= [(-1,-1), (-1,1), (1,-1), (1,1)]

color = np.zeros((window.height,window.width,4),np.ubyte).view(gloo.Texture2D)
color.interpolation = gl.GL_LINEAR
pick = np.zeros((window.height,window.width,4),np.ubyte).view(gloo.Texture2D)
pick.interpolation = gl.GL_LINEAR
framebuffer = gloo.FrameBuffer(color=[color,pick])
quad["color"] = color

index = 0
mouse = 0,0

@window.event
def on_init():
    pass
    # gl.glEnable(gl.GL_DEPTH_TEST)

@window.event
def on_draw(dt):
    gl.glEnable(gl.GL_DEPTH_TEST)

    framebuffer.activate()
    window.clear()
    program.draw(gl.GL_POINTS)
    if mouse is not None:
        gl.glReadBuffer(gl.GL_COLOR_ATTACHMENT1, gl.GL_FRONT)
        r,g,b,a = gl.glReadPixels(mouse[0],mouse[1],1,1, gl.GL_RGBA, gl.GL_UNSIGNED_BYTE)
        print('mouse {}'.format(mouse))
        if type(r) is not int: r = ord(r)
        if type(g) is not int: g = ord(g)
        if type(b) is not int: b = ord(b)
        print('r {} g {} b {} a {}'.format(r, g, b, a))
        index = b + 256*g + 256*256*r

        if index < len(program):
            program["bg_color"][index] = 0,0,0,1
    framebuffer.deactivate()
    gl.glDisable(gl.GL_DEPTH_TEST)  
    quad.draw(gl.GL_TRIANGLE_STRIP)

@window.event
def on_resize(width, height):
    framebuffer.resize(width,height)
    quad["color"] = framebuffer.color[0]

@window.event
def on_mouse_motion(x,y, dx, dy):
    global mouse
    mouse = int(x), window.height-int(y)

window.attach(program['transform'])
app.run()

```

**Comment by rougier** [comment]
Could you use a specific seed for the random generator (let's say np.random.seed(1)) and points me to the error you refer to (because I'm not sure I get it correctly). And maybe a lesser number of points could help the debugging.

**Comment by tobyclh** [comment]
No problem, I set np.random.seed(1) and number of points to 10
![first](https://user-images.githubusercontent.com/12501995/36629130-3a637ac2-1995-11e8-83ff-35ad65f85fa2.jpeg)
![second](https://user-images.githubusercontent.com/12501995/36629131-3a8b1348-1995-11e8-9913-bb4ad52ab8b7.jpeg)
![third](https://user-images.githubusercontent.com/12501995/36629132-3ab26be6-1995-11e8-81f1-264c9ffc2953.jpeg)

Notice how the black dot is moving right, while both pinkish(dot that overlaps with the black dot at the second image) and skin-color dots(overlaps with the black dot at the third image) are moving left.
Nonetheless one is on top and the other is below the black dot.

**Comment by rougier** [comment]
I 'm not sure I see the problem. Here is a slightly modified version. I got rid of picking, put discs on the same radius and colored them according to depth. It seems to be ok (do you confirm?) and I see no reason why it would be different with varying radius.


![debug-glumpy](https://user-images.githubusercontent.com/327203/36630829-6354cd32-196d-11e8-8c65-2941b996078b.png)

```
import numpy as np
from glumpy import app, gl, gloo, glm
from glumpy.transforms import Trackball

vertex = """
#include "math/constants.glsl"
uniform float linewidth;
uniform float antialias;
attribute float size;
attribute vec3  position;
varying float v_size, v_depth;
void main (void)
{
    v_size = size;
    gl_Position = <transform(vec4(position,1.0))>;
    v_depth = gl_Position.z / gl_Position.w;
    gl_PointSize = M_SQRT2 * size + 2.0 * (linewidth + 1.5*antialias);
}
"""

fragment = """
#include "markers/disc.glsl"
#include "math/constants.glsl"
#include "antialias/antialias.glsl"
uniform float linewidth;
uniform float antialias;
varying float v_size, v_depth;
void main()
{
    vec2 P = gl_PointCoord.xy - vec2(0.5,0.5);
    float point_size = M_SQRT2*v_size  + 2 * (linewidth + 1.5*antialias);
    float distance = marker_disc(P*point_size, v_size);
    vec4 fg_color = vec4(0,0,0,1);
    vec4 bg_color = vec4(vec3(1-v_depth),1);
    gl_FragColor = outline(distance, linewidth, antialias, fg_color, bg_color);
}
"""

window = app.Window(width=1024, height=1024, color=(1,1,1,1))

np.random.seed(1)
n = 1000

program = gloo.Program(vertex, fragment, count=n)
P = np.random.normal(0,1,(n,3))
P /= np.sqrt((P**2).sum(axis=1)).reshape(n,1)
program['position'] = P
program['size'] = np.random.uniform(30,40,n)
program['linewidth'] = 1.0
program['antialias'] = 1.0
program['transform'] = Trackball(zfar=5, distance=4)

@window.event
def on_init():
    gl.glEnable(gl.GL_DEPTH_TEST)

@window.event
def on_draw(dt):
    window.clear()
    program.draw(gl.GL_POINTS)

window.attach(program['transform'])
app.run()
```

**Comment by tobyclh** [comment]
Thanks for putting up this nice script, 
I run the script and got the same result you got.
So it seems like the problem is introduced by picking.
As I modify your script, but add back the picking, this time the depth test is clearly messed up.
![image](https://user-images.githubusercontent.com/12501995/36631158-5ab6ae9a-19b6-11e8-9ec4-7ba121212d8b.png)

Below is the script 
```

import numpy as np
from glumpy import app, gl, gloo, glm
from glumpy.transforms import Trackball
from OpenGL.GL import * 
app.use('qt5')
np.random.seed(1)
vertex = """
#include "math/constants.glsl"

uniform float linewidth;
uniform float antialias;

attribute vec4  fg_color;
attribute vec4  bg_color;
attribute float size;
attribute vec3  position;
attribute float id;

varying vec4  v_id;

varying float v_size, v_depth;
void main (void)
{
    // id to vec3 (alpha must be one or color will be blended
    v_id = vec4 ( mod(floor(id / (256*256)), 256) / 255.0,
                  mod(floor(id /     (256)), 256) / 255.0,
                  mod(floor(id /       (1)), 256) / 255.0,
                  1.0 );

    v_size = size;
    gl_Position = <transform(vec4(position,1.0))>;
    v_depth = gl_Position.z / gl_Position.w;
    gl_PointSize = M_SQRT2 * size + 2.0 * (linewidth + 1.5*antialias);
}
"""

fragment = """
#include "markers/disc.glsl"
#include "math/constants.glsl"
#include "antialias/antialias.glsl"

uniform float linewidth;
uniform float antialias;

varying vec4  v_id;
varying float v_size, v_depth;

void main()
{
    vec2 P = gl_PointCoord.xy - vec2(0.5,0.5);
    float point_size = M_SQRT2*v_size  + 2 * (linewidth + 1.5*antialias);
    float distance = marker_disc(P*point_size, v_size);

    vec4 color = outline(distance, linewidth, antialias, vec4(0,0,0,1), vec4(vec3(1-v_depth),1));

    gl_FragData[0] = color;
    gl_FragData[1] = v_id;
}
"""

quad_vertex = """
attribute vec2 position;
varying vec2 v_texcoord;
void main(void)
{
    gl_Position = vec4(position,0,1);
    v_texcoord = (position+1.0)/2.0;
}"""

quad_fragment = """
uniform sampler2D color;
varying vec2 v_texcoord;
void main(void)
{
    gl_FragColor = texture2D(color,v_texcoord);
}
"""

window = app.Window(width=1000, height=1000, color=(1,1,1,1))

n = 5000
program = gloo.Program(vertex, fragment, count=n)
program['position'] = 0.75 * np.random.randn(n,3)
program['size']     = np.random.uniform(20,30,n)
program['fg_color'] = 0,0,0,1
program['bg_color'] = np.random.uniform(0.75, 1.00, (n, 4))
program['bg_color'][:,3] = 1
program['linewidth'] = 1.0
program['antialias'] = 1.0
program['transform'] = Trackball()
program["id"] = np.arange(n,dtype=np.float32)

quad = gloo.Program(quad_vertex, quad_fragment, count=4)
quad['position']= [(-1,-1), (-1,1), (1,-1), (1,1)]

color = np.zeros((window.height,window.width,4),np.ubyte).view(gloo.Texture2D)
color.interpolation = gl.GL_LINEAR
pick = np.zeros((window.height,window.width,4),np.ubyte).view(gloo.Texture2D)
pick.interpolation = gl.GL_LINEAR
framebuffer = gloo.FrameBuffer(color=[color,pick])
quad["color"] = color

index = 0
mouse = 0,0

@window.event
def on_init():
    pass
    # gl.glEnable(gl.GL_DEPTH_TEST)

@window.event
def on_draw(dt):
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)
    # gl.glEnable(gl.GL_DEPTH_TEST)
    framebuffer.activate()
    window.clear()
    program.draw(gl.GL_POINTS)
    if mouse is not None:
        gl.glReadBuffer(gl.GL_COLOR_ATTACHMENT1)
        r,g,b,a = gl.glReadPixels(mouse[0],mouse[1],1,1, gl.GL_RGBA, gl.GL_UNSIGNED_BYTE)
        print('mouse {}'.format(mouse))
        if type(r) is not int: r = ord(r)
        if type(g) is not int: g = ord(g)
        if type(b) is not int: b = ord(b)
        print('r {} g {} b {} a {}'.format(r, g, b, a))
        index = b + 256*g + 256*256*r

        if index < len(program):
            program["bg_color"][index] = 0,0,0,1
    framebuffer.deactivate()
    # gl.glDisable(gl.GL_DEPTH_TEST)  
    quad.draw(gl.GL_TRIANGLE_STRIP)

@window.event
def on_resize(width, height):
    framebuffer.resize(width,height)
    quad["color"] = framebuffer.color[0]

@window.event
def on_mouse_motion(x,y, dx, dy):
    global mouse
    mouse = int(x), window.height-int(y)

window.attach(program['transform'])
app.run()

```

**Comment by rougier** [comment]
I see. I think the reason is the framebuffer has no depth buffer. You shoud add one and hopefully it will be working as expected (can you make a PR if it fixes the bug?)

**Comment by tobyclh** [comment]
I can confirm that its working with the PR :beers:
